ProcessLib/UncoupledProcessesTimeLoop.cpp

```diff
@@ -405,35 +459,98 @@ bool solveOneTimeStepOneProcess(GlobalVector& x, std::size_t const timestep,
     time_disc.pushState(t, x, mat_strg);
 
-    process.postTimestep(x);
-
     return nonlinear_solver_succeeded;
 }
 
 UncoupledProcessesTimeLoop::UncoupledProcessesTimeLoop(
     std::unique_ptr<NumLib::ITimeStepAlgorithm>&& timestepper,
     std::unique_ptr<Output>&& output,
-    std::vector<std::unique_ptr<SingleProcessData>>&& per_process_data)
+    std::vector<std::unique_ptr<SingleProcessData>>&& per_process_data,
+    const unsigned global_coupling_max_iterations,
+    std::unique_ptr<NumLib::ConvergenceCriterion>&& glb_coupling_conv_crit)
     : _timestepper{std::move(timestepper)},
       _output(std::move(output)),
-      _per_process_data(std::move(per_process_data))
+      _per_process_data(std::move(per_process_data)),
+      _global_coupling_max_iterations(global_coupling_max_iterations),
+      _global_coupling_conv_crit(std::move(glb_coupling_conv_crit))
 {
 }
 
+/**
+ *  This function fills the vector of solutions of coupled processes of
+ *  processes, _solutions_of_coupled_processes, and initializes the vector of
+ *  solutions of the previous coupling iteration,
+ *  _solutions_of_last_cpl_iteration.
+ *
+ *  \return a boolean value as a flag to indicate there should be a coupling
+ *          among process or not.
+ */
+bool UncoupledProcessesTimeLoop::setCoupledSolutions()
+{
+    if (!_global_coupling_conv_crit && _global_coupling_max_iterations == 1)
+        return false;
+
+    unsigned pcs_idx = 0;
+    _solutions_of_coupled_processes.reserve(_per_process_data.size());
+    for (auto& spd : _per_process_data)
+    {
+        BaseLib::RunTime time_timestep_process;
+        time_timestep_process.start();
+
+        auto const& coupled_processes = spd->coupled_processes;
+        std::map<ProcessType, GlobalVector const&> coupled_xs;
+        auto it = coupled_processes.begin();
+        while (it != coupled_processes.end())
+        {
+            ProcessLib::Process const& coupled_process = it->second;
+            auto const found_item = std::find_if(
+                _per_process_data.begin(),
+                _per_process_data.end(),
+                [&coupled_process](
+                    std::unique_ptr<SingleProcessData> const& item) {
+                    return coupled_process.getProcessType() ==
+                           item->process.getProcessType();
+                });
+
+            if (found_item != _per_process_data.end())
+            {
+                // Id of the coupled process:
+                const std::size_t c_id = found_item - _per_process_data.begin();
+
+                BaseLib::insertMapIfKeyUniqueElseError(
+                    coupled_xs, it->first, *_process_solutions[c_id],
+                    "global_coupled_x");
+            }
+            it++;
+        }
+        _solutions_of_coupled_processes.emplace_back(coupled_xs);
+
+        const auto x = _process_solutions[pcs_idx];
+
+        auto x1 = &NumLib::GlobalVectorProvider::provider.getVector(*x);
```

Comment:
Can you give the variable `x1` a more sensible name, please, or explain it in a comment....